### DISCUSSION TITLE
How can i use hCaptcha verification to protect a table in supabase?

### DISCUSSION THREAD

**Discussion by ghost** [supabase/supabase#14403] [Questions]
In supabase: I have a table named "ratings", this table have three columns named; "rating", "review" and "name", the "ratings" table is basically there to store the rating info.

The problem: I want to grant insert access to all the users (not necessarily authenticated/logged in) who have successfully completed the hCpatcha challenge. i do not know the right RLS (Row Level Security) policy and database function to do so.

i tried the following function with the help of the supabase ai:
`
CREATE OR REPLACE FUNCTION hcaptcha_verified()
RETURNS boolean AS $$
DECLARE
  response text;
  api_response json;
BEGIN
  -- Get the hCaptcha response from the request headers
  response := (SELECT headers['x-hcaptcha-response'] FROM jsonb_to_record(supabase.headers()) AS headers);
  RAISE NOTICE 'hCaptcha response: %', response;
  -- Verify the hCaptcha response using the hCaptcha API
  api_response := http_post('https://hcaptcha.com/siteverify', 'application/x-www-form-urlencoded', 'secret=0xee72qwefga9f02E0c2178oqldfkwe0bb554f39EC&response=' || response)::json;
  RETURN api_response->>'success'::boolean;
END;
$$ LANGUAGE plpgsql;
`
and tried the following RLS policy along with that function:
`
hcaptcha_verified()
`

Apparently the problem is inside the `hcpatcha_verified()` function, in this function, the `supabase.headers()` is not the correct way to receive the sent headers,

any help would be appreciated.

**Comment by GaryAustin1** [comment]
IF (and I know nothing about it) hCaptcha sends that header with the supabase API request, then you read the header using this method:
https://postgrest.org/en/stable/references/transactions.html?highlight=headers#request-headers-cookies-and-jwt-claims
`SELECT current_setting('request.headers', true)::json->>'x-hcaptcha-response';`

You could look at the API log to see if that header is being sent, I believe.
